code/modules/food_and_drinks/drinks/drinks/cans.dm

```diff
@@ -1,17 +1,45 @@
 /obj/item/reagent_containers/food/drinks/cans
-	var/canopened = 0
+	var/canopened = FALSE
 	var/is_glass = 0
 	var/is_plastic = 0
+	var/shaken = 0
+	var/bursting = FALSE
+	var/burstchance = 0
 
 /obj/item/reagent_containers/food/drinks/cans/New()
 	..()
 	flags &= ~OPENCONTAINER
 
-/obj/item/reagent_containers/food/drinks/cans/attack_self(mob/user)
-	if(canopened == 0)
+/obj/item/reagent_containers/food/drinks/cans/attack_self(mob/user, burstopen = FALSE)
+	if(!canopened && shaken)
+		if(!burstopen)
+			to_chat(user, "<span class='notice'>You open the drink with an audible pop!</span>")
+		else
+			visible_message("<span class='warning'>The [name] bursts open!</span>")
+		if(shaken < 5)
+			to_chat(user, "<span class='warning'>The [name] fizzes violently!</span>")
+		else
+			to_chat(user, "<span class='warning'>The [name] erupts into foam!</span>")
+			foam_up()
+
+		for(var/mob/living/carbon/C in range(1, get_turf(src)))
+			C.visible_message("<span class='warning'>You are splattered with [name]!</span>")
+			reagents.reaction(C, REAGENT_TOUCH)
+			if(shaken > C.wetlevel)
+				C.wetlevel = shaken
+			else if(shaken == C.wetlevel)
+				C.wetlevel++
+
+		reagents.remove_any((shaken/5) * reagents.total_volume)
+		playsound(loc,'sound/effects/canopenfizz.ogg', rand(10,50), 1)
+		canopened = TRUE
+		flags |= OPENCONTAINER
```

Comment:
Can extract the top 3 lines to one scope below and remove line 40, 42, 43

Reply:
I moved canopend and flags ahead of the branching to avoid needlessly repeating that code, but the reagents.remove_any needs to be after reagents.react and playsound is circumstantially unique.